server/parser/src/main/java/org/eclipse/lsp/cobol/core/hw/antlradapter/ProcedureDivisionAntlrAdapter.java

```diff
@@ -215,6 +236,40 @@ private ParserRuleContext parseSentence(Statement node) {
     return antlrParser.sentence();
   }
 
+  private ProcedureDivisionContext parseProcedureDivision(ProcedureDivision node) {
+    StringBuilder sb = new StringBuilder();
+    for (CstNode c : node.getChildren()) {
+      if (hasToken(c, node.getBodyStartToken())) {
+        break;
+      }
+      sb.append(c.toText());
+    }
+    String input = generatePrefix(charStream, findStartToken(node).get()) + sb;
+    CobolProcedureDivisionLexer antlrLexer =
+            new CobolProcedureDivisionLexer(CharStreams.fromString(input));
+    antlrLexer.removeErrorListeners();
+    antlrLexer.addErrorListener(errorListener);
+    CommonTokenStream tokens = new CommonTokenStream(antlrLexer);
+    CobolProcedureDivisionParser antlrParser = new CobolProcedureDivisionParser(tokens);
+    antlrParser.removeErrorListeners();
+    antlrParser.addErrorListener(errorListener);
+    antlrParser.setErrorHandler(errorStrategy);
+    antlrParser.addParseListener(treeListener);
+    return antlrParser.procedureDivision();
+  }
+
+  private boolean hasToken(CstNode node, Token token) {
+    if (node.equals(token)) {
+      return true;
+    }
+    for (CstNode c : node.getChildren()) {
```

Comment:
Stream?

Reply:
In this case, it could be a bit more complicated. I'm not sure how to fit recursion into the stream.